### PR TITLE
basic dimred

### PR DESCRIPTION
**Pull request by bomquangia**

### CODE REVIEW

File: walnut/dimred.py [hunk 1]
```diff
@@ -1,10 +1,213 @@
...
 
 class Dimred:
-    def __init__(self, dimred_folder, TextFile: FileIO):
-        self.__dir = dimred_folder
-        self.__TextFile = TextFile
-        self.__metalist = JSONFile(os.path.join(self.__dir, "meta"), self.__TextFile)
-    
-    # TODO: needs more APIs
\ No newline at end of file
+	def __init__(self, dimred_folder, TextFile: common.FileIO=common.TextFile):
```
Comment (JATC1024) [inline walnut/dimred.py:26]: `TextFile` is a *class*, not an *instance*. Should change to
```
from typing import Type

def __init__(self, dimred_folder, Type[common.FileIO[str]]=Type[common.TextFile]):
```

If you are wondering why is it `FileIO[str]` and not `FileIO`, please look at [commons.py](https://github.com/bioturing/bbrowser_python_studyio/blob/main/walnut/common.py)

File: walnut/dimred.py [hunk 2]
```diff
@@ -1,10 +1,213 @@
...
+
+		Args:
+			dimred_info (dict): A dictionary with obligatory keys: ['coords', 'name', 'id', 'size', 'param'], and optionally ['history'].
+			Will be validated with pydantic to ensure all required slots (coords/slides, history, param, etc)
+			id (_type_, optional): id of the new dimred to be written. Will be generated if not passed. 
+			Dimred_info.id will be forced to be equal to id.
+		"""
+		if id is None:
+			id = common.create_uuid()
+		else: # Make sure that user's input id not already exists
```
Comment (JATC1024) [inline walnut/dimred.py:75]: Why not just
```
elif id in self.data.keys():
    print("warn something")
    return
```

File: walnut/dimred.py [hunk 3]
```diff
@@ -1,10 +1,213 @@
...
+
+
+	def __repr__(self):
+		""""return an overview of all dimred, id, name, type, dim"""
+		shapes = self.__get_shape()
+		return repr(pd.DataFrame({'IDs':self.ids, 'Names': self.names, 'Shape': shapes}))
+	
+
+class SingleDimred:
+	def __init__(self, path, TextFile: common.FileIO=common.TextFile):
```
Comment (JATC1024) [inline walnut/dimred.py:152]: Should also change to `Type[common.FileIO[str]]`

File: walnut/dimred.py (hunk 1 above)
Comment (JATC1024) [reply walnut/dimred.py:26]: Yes, but remember to use `Type[common.TextFile]` instead of just `common.TextFile`

File: walnut/models/dimred.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
+from pydantic import BaseModel
+from typing import List, Optional
+
+class DimredData(BaseModel):
+	id: Optional[str]=None
+	name: str
+	size: List[int]
+	history: List[dict]=None
```
Comment (JATC1024) [inline walnut/models/dimred.py:8]: Please consider using [this model](https://github.com/bioturing/bbrowser_python_studyio/blob/1c4ec374c762e9c0d2c9c597b1c29b6075953f5b/walnut/models/history.py) for `history`

Comment (bomquangia) [reply walnut/models/dimred.py]: Solved with https://github.com/bioturing/bbrowser_python_studyio/pull/3/commits/1212252831618f86986e360c32a6cd4756bdc6d6

**Pull request closed without merging**